### PR TITLE
vim-patch:8.1.{899,903,905,907,908,910,913,1746}

### LINKED ISSUE #9930: Crash when searching for column with number between 2^64 - 1 and 2 ^ 64 - 2 ^ 31
NVIM v0.4.0-dev
VIM - Vi IMproved 8.1 compiled from source today does not crash.
- Operating system/version: Ubuntu 18.10
- Terminal name/version: st
- `$TERM`: xterm-256color and st-256color both reproduce the problem

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
/\v%18446744071562067968c
```
Will yield: 
```
nvim: /build/neovim-5zUQcT/neovim-0.4.0+ubuntu1+git201904202022-43356a4-eada8be/src/nvim/regexp_nfa.c:5926: nfa_regmatch: Assertion `t->state->val >= 0 && reginput >= regline && (uintmax_t)(reginput - regline) <= UINTMAX_MAX - 1' failed.
                                               
[1]    30948 abort (core dumped)  /usr/bin/nvim -u NORC
```

Any number between 18446744071562067968 (2^64 - 2^31) and 18446744073709551615 (2^64 - 1) causes a crash for me. The issue doesn't reproduce for 2^64 - 2^32 but I am not sure where the issue stops occuring between 2^64 - 2^31 and 2^64 - 2^32.

I ran into this while generating patterns using c++ and overflowing a size_t value. I can't imagine this is a problem for anyone, but it does seem at least to me like something that should get fixed if possible.

### PR DESCRIPTION
**Pull request by janlazo**
Close https://github.com/neovim/neovim/pull/10051
Close #9930.

Thanks @gafeol.

### CODE REVIEW

**Comment by justinmk** [comment]
should we use @jamessan 's suggestion in https://github.com/neovim/neovim/pull/10051#issuecomment-500223981 ?

Does this close https://github.com/neovim/neovim/issues/9930 ?

File: src/nvim/regexp_nfa.c [hunk 1]
```diff
@@ -1493,8 +1493,8 @@ static int nfa_regatom(void)
 
     default:
     {
-      long n = 0;
-      int cmp = c;
+      uint64_t n = 0;
+      uint32_t cmp = c;
```
Comment (janlazo) [inline src/nvim/regexp_nfa.c]: Do I need an assertion for this? `c` is int.

Comment (justinmk) [reply src/nvim/regexp_nfa.c]: > refactoring the other regexp functions to return uint32_t will take a while. 

Just an assert is fine. Refactoring regexp module is out of scope, not worth our time.

Comment (janlazo) [reply src/nvim/regexp_nfa.c]: Actually, does `cmp` need to be fixed? It doesn't impact `n`.

File: src/nvim/regexp_nfa.c [hunk 2]
```diff
@@ -1493,16 +1493,22 @@ static int nfa_regatom(void)
 
     default:
     {
-      long n = 0;
-      int cmp = c;
+      int64_t n = 0;
```
Comment (janlazo) [inline src/nvim/regexp_nfa.c]: This has to be a signed type to not trigger the assertion in the new test.

File: test/functional/eval/match_functions_spec.lua [hunk 3]
```diff
@@ -156,3 +156,12 @@ describe('matchaddpos()', function()
     ]], {[1] = {foreground = Screen.colors.Red}, [2] = {bold = true, foreground = Screen.colors.Blue1}})
   end)
 end)
+
+describe('nfa_regatom() column search', function()
```
Comment (justinmk) [inline test/functional/eval/match_functions_spec.lua]: In the merge, I moved this test to a new test file `normal/search_spec.lua`

**Pull request merged**